File: clients/hermes/python/src/membase_hermes/spool.py
"""Failure-path capture spool for Hermes (ADR 0005 / DR-2).

Python port of the TS @membase/capture-core disk spool. The normal capture
path stays in-memory (capture.py's worker queue) — ADR 0002's "don't force the
disk abstraction on the long-lived host". This spool is used ONLY when a live
upload fails after retries: instead of dropping the batch (lost forever), it is
written to disk so a later `hermes-membase dream` (or a startup drain) uploads
it.

The on-disk contract — record schema and `capture_id` hashing — is bound to the
TS spool by golden vectors (spec/spool-vectors.json), so the two languages can't
silently drift (same mechanism as sanitize/handoff, ADR 0002). The file-locking
and retry mechanics are per-language by design and are not vectorized.
"""

from __future__ import annotations

import hashlib
import json
import os
import time
from collections.abc import Callable
from pathlib import Path
from typing import Any

MIN_CONTENT_LENGTH = 20
LOCK_WAIT_MS = 2000
LOCK_STALE_MS = 30_000
INFLIGHT_STALE_MS = 60_000
SENT_LEDGER_MAX = 2000
DISPLAY_SUMMARY_MAX = 180


def _sha256_hex(value: str) -> str:
    return hashlib.sha256(value.encode("utf-8")).hexdigest()


def _truncate(text: str, limit: int) -> str:
    return text if len(text) <= limit else text[: max(0, limit - 1)] + "…"


def compute_capture_id(
    *, session_id: str | None, capture_kind: str, sanitized_content: str
) -> str:
    """Matches the TS captureId: sha256("{session}:{kind}:{sanitized}").

    Content is ALREADY sanitized by the caller — the TS side sanitizes inside
    captureId, but enqueue() here sanitizes once up front and reuses the result,
    which yields the identical hash input. Bound by spool-vectors.json.
    """
    session = session_id or "unknown"
    return _sha256_hex(f"{session}:{capture_kind}:{sanitized_content}")


class CaptureSpool:
    """Disk-backed spool rooted at ``<state_dir>/spool``."""

    def __init__(
        self, state_dir: Callable[[], Path], sanitize: Callable[[str], str]
    ) -> None:
        self._state_dir = state_dir
        self._sanitize = sanitize

    # --- paths -----------------------------------------------------------
    def _spool_dir(self) -> Path:
        d = self._state_dir() / "spool"
        d.mkdir(parents=True, exist_ok=True, mode=0o700)
        return d

    def _pending_path(self) -> Path:
        return self._spool_dir() / "pending.jsonl"

    def _sent_path(self) -> Path:
        return self._spool_dir() / "sent.json"

    def _lock_path(self) -> Path:
        return self._spool_dir() / ".lock"

    def _inflight_path(self) -> Path:
        return self._spool_dir() / f"inflight-{os.getpid()}-{int(time.time() * 1000)}.jsonl"

    def _inflight_files(self) -> list[Path]:
        return sorted(
            p
            for p in self._spool_dir().iterdir()
            if p.name.startswith("inflight-") and p.suffix == ".jsonl"
        )

    # --- locking ---------------------------------------------------------
    def _acquire_lock(self) -> int:
        path = self._lock_path()
        deadline = time.monotonic() + LOCK_WAIT_MS / 1000
        while time.monotonic() < deadline:
            try:
                return os.open(path, os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o600)
            except FileExistsError:
                try:
                    age_ms = (time.time() - path.stat().st_mtime) * 1000
                    if age_ms > LOCK_STALE_MS:
                        path.unlink(missing_ok=True)
                        continue
                except FileNotFoundError:
                    continue
                time.sleep(0.025)
        raise TimeoutError("Timed out waiting for Membase capture spool lock.")

    def _release_lock(self, fd: int) -> None:
        try:
            os.close(fd)
        except OSError:
            pass
        self._lock_path().unlink(missing_ok=True)

    def _with_lock(self, fn: Callable[[], Any]) -> Any:
        fd = self._acquire_lock()
        try:
            return fn()
        finally:
            self._release_lock(fd)

    # --- record io -------------------------------------------------------
    def _read_records_from(self, path: Path) -> list[dict[str, Any]]:
        if not path.exists():
            return []
        out: list[dict[str, Any]] = []
        for line in path.read_text("utf-8").splitlines():
            line = line.strip()
            if not line:
                continue
            try:
                out.append(json.loads(line))
            except json.JSONDecodeError:
                continue
        return out

    def _read_records(self) -> list[dict[str, Any]]:
        return self._read_records_from(self._pending_path())

    def _write_records_to(self, path: Path, records: list[dict[str, Any]]) -> None:
        tmp = path.with_suffix(path.suffix + ".tmp")
        body = "".join(json.dumps(r) + "\n" for r in records)
        tmp.write_text(body, "utf-8")
        os.chmod(tmp, 0o600)
        os.replace(tmp, path)

    def _write_records(self, records: list[dict[str, Any]]) -> None:
        self._write_records_to(self._pending_path(), records)

    def _recover_stale_inflight(self) -> None:
        """Re-queue records from crashed flushes.

        A flush claims its batch into an ``inflight-*.jsonl`` file (see flush);
        if the process dies before the batch is requeued, the file lingers.
        Any inflight file older than INFLIGHT_STALE_MS is assumed abandoned and
        its records are appended back to pending. Callers must hold the lock.
        Mirrors the TS spool's recoverStaleInflightLocked.
        """
        now_ms = time.time() * 1000
        for path in self._inflight_files():
            try:
                if now_ms - path.stat().st_mtime * 1000 < INFLIGHT_STALE_MS:
                    continue
                stale = self._read_records_from(path)
                if stale:
                    self._write_records([*self._read_records(), *stale])
                path.unlink(missing_ok=True)
            except FileNotFoundError:
                continue

    def _read_sent_ids(self) -> set[str]:
        path = self._sent_path()
        if not path.exists():
            return set()
        try:
            data = json.loads(path.read_text("utf-8"))
            return set(data) if isinstance(data, list) else set()
        except (json.JSONDecodeError, OSError):
            return set()

    def _write_sent_ids(self, ids: set[str]) -> None:
        # Cap the ledger like the TS spool's writeSentIds slice(-2000): the
        # ledger only guards against re-upload of records still on disk, so an
        # unbounded set would grow (and be rewritten O(n)) on every mark.
        path = self._sent_path()
        tmp = path.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(sorted(ids)[-SENT_LEDGER_MAX:]), "utf-8")
        os.chmod(tmp, 0o600)
        os.replace(tmp, path)

    # --- public API ------------------------------------------------------
    def enqueue_capture(
        self,
        *,
        content: str,
        capture_kind: str = "conversation",
        session_id: str | None = None,
        display_summary: str | None = None,
        project: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any] | None:
        """Persist one capture; returns the record, or None if dropped.

        Drops (returns None) when the sanitized content is too short or is a
        duplicate of a pending or already-sent record — matching the TS spool.
        """
        sanitized = self._sanitize(content)
        if not sanitized or len(sanitized) < MIN_CONTENT_LENGTH:
            return None
        capture_id = compute_capture_id(
            session_id=session_id,
            capture_kind=capture_kind,
            sanitized_content=sanitized,
        )
        record: dict[str, Any] = {
            "capture_id": capture_id,
            "capture_kind": capture_kind,
            "content": sanitized,
            "display_summary": display_summary
            or _truncate(sanitized, DISPLAY_SUMMARY_MAX),
            "project": project,
            "metadata": metadata or {},
            "created_at": _iso_now(),
            "attempts": 0,
        }

        def _do() -> dict[str, Any] | None:
            self._recover_stale_inflight()
            existing = self._read_records()
            if any(r.get("capture_id") == capture_id for r in existing):
                return None
            if capture_id in self._read_sent_ids():
                return None
            self._write_records([*existing, record])
            return record

        try:
            return self._with_lock(_do)
        except TimeoutError:
            return None

    def pending_count(self) -> int:
        def _count() -> int:
            self._recover_stale_inflight()
            return len(self._read_records())

        return self._with_lock(_count)

    def flush(
        self, send: Callable[[dict[str, Any]], None], limit: int = 50
    ) -> tuple[int, int]:
        """Upload up to ``limit`` records via ``send`` (which raises on failure).

        Returns (flushed, remaining). A record whose send raises is kept for a
        later dream. Safe to call repeatedly; the sent ledger blocks re-upload.
        """

        inflight = self._inflight_path()

        def _claim() -> list[dict[str, Any]]:
            self._recover_stale_inflight()
            sent = self._read_sent_ids()
            records = [
                r for r in self._read_records() if r.get("capture_id") not in sent
            ]
            batch = records[:limit]
            self._write_records(records[limit:])
            # Persist the claimed batch so a crash mid-send can't lose it: the
            # records leave pending.jsonl only after landing in an inflight
            # file, which _recover_stale_inflight sweeps back if we die.
            if batch:
                self._write_records_to(inflight, batch)
            return batch

        batch = self._with_lock(_claim)
        if not batch:
            return (0, self.pending_count())

        failed: list[dict[str, Any]] = []
        flushed = 0
        for record in batch:
            try:
                send(record)
            except Exception as error:  # noqa: BLE001 - uploader errors are data
                failed.append(
                    {
                        **record,
                        "attempts": int(record.get("attempts", 0)) + 1,
                        "last_error": str(error),
                    }
                )
                continue

            def _mark_sent(rid: str = record["capture_id"]) -> None:
                sent = self._read_sent_ids()
                sent.add(rid)
                self._write_sent_ids(sent)

            self._with_lock(_mark_sent)
            flushed += 1

        def _requeue() -> int:
            if failed:
                self._write_records([*self._read_records(), *failed])
            inflight.unlink(missing_ok=True)
            return len(self._read_records())

        remaining = self._with_lock(_requeue)
        return (flushed, remaining)


def _iso_now() -> str:
    # ISO 8601 with a trailing Z, matching the TS `new Date().toISOString()`.
    # Both parts come from ONE clock read so a second-rollover between them
    # can't skew the seconds and milliseconds apart.
    now = time.time()
    return (
        time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime(now))
        + f".{int((now % 1) * 1000):03d}Z"
    )


def default_capture_spool() -> CaptureSpool:
    """The process-wide spool rooted at the Hermes home (~/.hermes)."""
    from .config import get_hermes_home
    from .sanitize import sanitize_capture_text

    return CaptureSpool(
        state_dir=get_hermes_home,
        sanitize=sanitize_capture_text,
    )

File: packages/capture-core/src/spool.ts
// Disk-persisted capture spool.
//
// Extracted from the Claude runtime. This design exists for hosts whose hook
// handlers are SHORT-LIVED SPAWNED PROCESSES (Claude Code today, Cursor's
// hooks.json processes): captures must survive across invocations, so
// the queue lives on disk with a lock file, crash-safe inflight handoff, and
// a sent-id ledger for dedupe. Long-lived hosts (OpenClaw gateway, Hermes)
// keep their in-process queues — this is deliberately NOT one abstraction
// over both models.
import { createHash } from "node:crypto";
import {
  appendFileSync,
  closeSync,
  existsSync,
  mkdirSync,
  openSync,
  readFileSync,
  readdirSync,
  rmSync,
  statSync,
} from "node:fs";
import { join } from "node:path";
import { truncateText } from "./index.js";
import { writeTextAtomic } from "./token-store.js";

const LOCK_STALE_MS = 30_000;
const LOCK_WAIT_MS = 2_000;
const INFLIGHT_STALE_MS = 60_000;
const SLEEP_BUFFER = new SharedArrayBuffer(4);
const SLEEP_VIEW = new Int32Array(SLEEP_BUFFER);

export interface SpoolRecord {
  capture_id: string;
  capture_kind: string;
  content: string;
  display_summary?: string;
  project?: string;
  metadata: Record<string, unknown>;
  created_at: string;
  attempts?: number;
  last_error?: string;
}

export interface CaptureSpoolOptions {
  /** Parent state directory; a `spool/` subdir is created inside (0700). */
  stateDir: () => string;
  /** Host sanitizer applied to content before hashing/storing. */
  sanitize: (text: string) => string;
  /** Captures shorter than this (after sanitize) are dropped. */
  minContentLength?: number;
}

export interface CaptureSpool {
  captureId(args: {
    sessionId?: string;
    captureKind: string;
    content: string;
  }): string;
  enqueueCapture(
    record: Omit<SpoolRecord, "created_at" | "capture_id"> & {
      sessionId?: string;
    },
  ): SpoolRecord | null;
  flushSpool(
    // biome-ignore lint/suspicious/noConfusingVoidType: senders may legitimately return nothing; undefined would force every caller to return a value
    send: (record: SpoolRecord) => Promise<void | boolean>,
    limit?: number,
  ): Promise<{ flushed: number; remaining: number }>;
  pendingSpoolCount(): number;
}

function sleepSync(ms: number): void {
  Atomics.wait(SLEEP_VIEW, 0, 0, ms);
}

function hash(input: string): string {
  return createHash("sha256").update(input).digest("hex");
}

export function createCaptureSpool(
  options: CaptureSpoolOptions,
): CaptureSpool {
  const minContentLength = options.minContentLength ?? 20;

  function spoolDir(): string {
    const dir = join(options.stateDir(), "spool");
    mkdirSync(dir, { recursive: true, mode: 0o700 });
    return dir;
  }

  function spoolPath(): string {
    return join(spoolDir(), "pending.jsonl");
  }

  function sentPath(): string {
    return join(spoolDir(), "sent.json");
  }

  function lockPath(): string {
    return join(spoolDir(), ".lock");
  }

  function inflightPath(): string {
    return join(spoolDir(), `inflight-${process.pid}-${Date.now()}.jsonl`);
  }

  function acquireLock(timeoutMs = LOCK_WAIT_MS): () => void {
    const path = lockPath();
    const deadline = Date.now() + timeoutMs;
    while (Date.now() < deadline) {
      try {
        const fd = openSync(path, "wx", 0o600);
        return () => {
          try {
            closeSync(fd);
          } catch {}
          try {
            rmSync(path, { force: true });
          } catch {}
        };
      } catch (error) {
        if ((error as NodeJS.ErrnoException).code !== "EEXIST") throw error;
        try {
          if (Date.now() - statSync(path).mtimeMs > LOCK_STALE_MS) {
            rmSync(path, { force: true });
            continue;
          }
        } catch {}
        sleepSync(25);
      }
    }
    throw new Error("Timed out waiting for Membase capture spool lock.");
  }

  function withSpoolLock<T>(callback: () => T, timeoutMs = LOCK_WAIT_MS): T {
    const release = acquireLock(timeoutMs);
    try {
      return callback();
    } finally {
      release();
    }
  }

  function captureId(args: {
    sessionId?: string;
    captureKind: string;
    content: string;
  }): string {
    return hash(
      `${args.sessionId ?? "unknown"}:${args.captureKind}:${options.sanitize(
        args.content,
      )}`,
    );
  }

  function readRecordsFromPath(path: string): SpoolRecord[] {
    if (!existsSync(path)) return [];
    const raw = readFileSync(path, "utf-8").trim();
    if (!raw) return [];
    return raw
      .split(/\r?\n/)
      .map((line) => {
        try {
          return JSON.parse(line) as SpoolRecord;
        } catch {
          return null;
        }
      })
      .filter((record): record is SpoolRecord => Boolean(record));
  }

  function readRecords(): SpoolRecord[] {
    return readRecordsFromPath(spoolPath());
  }

  function writeRecordsToPath(path: string, records: SpoolRecord[]): void {
    writeTextAtomic(
      path,
      records.map((record) => JSON.stringify(record)).join("\n") +
        (records.length ? "\n" : ""),
    );
  }

  function writeRecords(records: SpoolRecord[]): void {
    writeRecordsToPath(spoolPath(), records);
  }

  function appendRecords(records: SpoolRecord[]): void {
    if (records.length === 0) return;
    appendFileSync(
      spoolPath(),
      `${records.map((record) => JSON.stringify(record)).join("\n")}\n`,
      {
        encoding: "utf-8",
        mode: 0o600,
      },
    );
  }

  function readSentIds(): Set<string> {
    const path = sentPath();
    if (!existsSync(path)) return new Set();
    try {
      const parsed = JSON.parse(readFileSync(path, "utf-8")) as unknown;
      if (!Array.isArray(parsed)) return new Set();
      return new Set(
        parsed.filter((value): value is string => typeof value === "string"),
      );
    } catch {
      return new Set();
    }
  }

  function writeSentIds(ids: Set<string>): void {
    const values = Array.from(ids).slice(-2000);
    writeTextAtomic(sentPath(), `${JSON.stringify(values, null, 2)}\n`);
  }

  function inflightFiles(): string[] {
    return readdirSync(spoolDir())
      .filter((name) => name.startsWith("inflight-") && name.endsWith(".jsonl"))
      .map((name) => join(spoolDir(), name));
  }

  function readInflightRecords(): SpoolRecord[] {
    return inflightFiles().flatMap((path) => readRecordsFromPath(path));
  }

  function dedupeRecords(
    records: SpoolRecord[],
    sentIds = readSentIds(),
  ): SpoolRecord[] {
    const seen = new Set<string>();
    return records.filter((record) => {
      if (sentIds.has(record.capture_id) || seen.has(record.capture_id)) {
        return false;
      }
      seen.add(record.capture_id);
      return true;
    });
  }

  function appendPendingRecordsLocked(records: SpoolRecord[]): void {
    const sentIds = readSentIds();
    const existingIds = new Set(
      readRecords().map((record) => record.capture_id),
    );
    const next = records.filter((record) => {
      if (sentIds.has(record.capture_id) || existingIds.has(record.capture_id)) {
        return false;
      }
      existingIds.add(record.capture_id);
      return true;
    });
    appendRecords(next);
  }

  function recoverStaleInflightLocked(): void {
    const now = Date.now();
    for (const path of inflightFiles()) {
      try {
        if (now - statSync(path).mtimeMs < INFLIGHT_STALE_MS) continue;
        appendPendingRecordsLocked(readRecordsFromPath(path));
        rmSync(path, { force: true });
      } catch {}
    }
  }

  function enqueueCapture(
    record: Omit<SpoolRecord, "created_at" | "capture_id"> & {
      sessionId?: string;
    },
  ): SpoolRecord | null {
    const content = options.sanitize(record.content);
    if (!content || content.length < minContentLength) return null;
    const next: SpoolRecord = {
      capture_id: captureId({
        sessionId: record.sessionId,
        captureKind: record.capture_kind,
        content,
      }),
      capture_kind: record.capture_kind,
      content,
      // Caller-supplied display_summary is raw hook/tool text — sanitize it
      // like content so secrets can't reach disk via the summary field.
      display_summary: record.display_summary
        ? options.sanitize(record.display_summary)
        : truncateText(content, 180),
      project: record.project,
      metadata: record.metadata,
      created_at: new Date().toISOString(),
      attempts: 0,
    };
    try {
      return withSpoolLock(() => {
        recoverStaleInflightLocked();
        const existing = [...readRecords(), ...readInflightRecords()];
        if (existing.some((item) => item.capture_id === next.capture_id)) {
          return null;
        }
        if (readSentIds().has(next.capture_id)) return null;
        appendRecords([next]);
        return next;
      });
    } catch {
      return null;
    }
  }

  function pendingSpoolCount(): number {
    return withSpoolLock(() => {
      recoverStaleInflightLocked();
      return readRecords().length;
    });
  }

  async function flushSpool(
    // biome-ignore lint/suspicious/noConfusingVoidType: senders may legitimately return nothing; undefined would force every caller to return a value
    send: (record: SpoolRecord) => Promise<void | boolean>,
    limit = 10,
  ): Promise<{ flushed: number; remaining: number }> {
    const drained = withSpoolLock(() => {
      recoverStaleInflightLocked();
      const sentIds = readSentIds();
      const records = dedupeRecords(readRecords(), sentIds);
      const batch = records.slice(0, limit);
      const pending = records.slice(limit);
      writeRecords(pending);
      const path = batch.length > 0 ? inflightPath() : undefined;
      if (path) writeRecordsToPath(path, batch);
      return { batch, path };
    });
    if (drained.batch.length === 0) {
      return { flushed: 0, remaining: pendingSpoolCount() };
    }

    const failed: SpoolRecord[] = [];
    let flushed = 0;
    for (const record of drained.batch) {
      try {
        // A resolved `false` counts as failure too — JS callers signalling
        // by return value must not silently drop the record.
        if ((await send(record)) === false) {
          throw new Error("uploader returned false");
        }
        withSpoolLock(() => {
          const sentIds = readSentIds();
          sentIds.add(record.capture_id);
          writeSentIds(sentIds);
        });
        flushed += 1;
      } catch (error) {
        failed.push({
          ...record,
          attempts: (record.attempts ?? 0) + 1,
          // Uploader errors can echo response bodies; sanitize and clamp
          // before persisting to disk.
          last_error: truncateText(
            options.sanitize(
              error instanceof Error ? error.message : String(error),
            ),
            300,
          ),
        });
      }
    }

    const remaining = withSpoolLock(() => {
      appendPendingRecordsLocked(failed);
      if (drained.path) rmSync(drained.path, { force: true });
      return readRecords().length;
    });
    return { flushed, remaining };
  }

  return { captureId, enqueueCapture, flushSpool, pendingSpoolCount };
}
